### ISSUE TITLE
Multiple simultaneous requests with variable threshold setup

### ISSUE THREAD

**Issue by digzol** [Yousei9/Logistic-Train-Network#146]
Instead of using the request threshold signal, I have a small setup of 3 combinators that send a signal requesting items that fall under a certain threshold. This way, I can make sure that train cargo slots are always roughly loaded equally, no matter the stack limit of the item being requested.

However, I come into a problem when the station buffer chest is empty; The station immediately makes several small requests that are all way below the item signal being sent. Lowering the threshold of stations from default 1,000 to 1 makes even less sense when deliveries of 86 items able to stack to 200 are made.

![image](https://i.imgur.com/LQP6ySD.png)

Scenario explained step by step:

- The content of the buffer chest is added to a negative of the maximum storage to hold (i.e. -20k for Sulfur, stacking to 200).
- The value is compared to a threshold value in a decider combinator (i.e. -14k) and outputs 1 since true.
- The output value is multiplied by the maximum load of a train (i.e -14k).
- Station receives signal of -14k for Sulfur, and a delivery of 14,000 is scheduled.
- The train loads the resources without problem and starts unloading the items back at the requesting station.
- **The issue:** for some reason, as soon as ~2.5k items are unloaded, the station simultaneously makes 3 new requests, typically ranging between 1k and 1.5k, if left to default station threshold of 1k.

The setup:

![image](https://i.imgur.com/6u12Gq6.png)

Additional testing:

- As mentioned, the new requests always happen when ~2.5k items are unloaded.
- At that moment, the signal for -14k items is still being emitted because (-20k + 2.5k < -14k).
- Happens no matter what providing station is used, providing threshold doesn't matter either.
- In this scenario, the final provided amount is 17k, with 600 items remaining at the providing station.
- Removing the setup fixes the issue.
- Adding a decider combinator that voids all signals if blue signal != 0 also fixes the issue and acts as a workaround.

I thought maybe the issue was on my side, afterall, the voiding signals when a train is at the station seems like a viable workaround. However, I can't see how doing 3+ simultaneous requests that could easily be merged into a single trip makes any sense. Especially while the first delivery fulfils the entirety of the signal being sent to LTN. At this point I'm going to assume it's a bug related to signal updates for now.

**EDIT:** The blueprint string of the setup: 
```
0eNrFVtuumzAQ/Jd9rOAIkzuq+hV9q46QA5vEkrGRsaMTRfx7bQiXJJBCVLUvkRzbs7Ozs4uvsOcGc8WEhugKLJGigOjXFQp2FJS7//QlR4iAaczAA0Ezt1KUcSg9YCLFL4hI+ekBCs00w/p+tbjEwmR7VPZAezPDlJnMR46JVizxc8nR4uaysJelcBEtYPCx8uACkb/4WNkwlpaw51nF7grE/ShM+5GYXYWld7e2GAlTiWH6tv1Zlr0jDbmwJcflkRXastI2QeEXWuY+E7nRzwx90lAMG4paSR7v8UTPTCp3rAlu99L27oGpQsdP+p6Z0sb+03KpT/hUXPSJiSPUMQpNXakCt8hyqqh2oeCH2zYF2lBcKiuSVganC0cehFuPKLV4qZQ0eliqsC/VAOzyFewQ3g1tRHaFNI0PSmZxhVKL4UG1iB1kjun0AvwESxms7jUDgIEEVu0lqpg+ZehSsAXaM1EV6CmFzjwjOXQ4nXuKN+yDNDk5bxToYOLOQT5ZBoG1kcxRNal9s9frIs4NUE732va5KT0IR405bMR1S6PJ6KXcXbMGo816YFyjGpl+T9lzLfzMOrT2KUdx1Ke6RY1Td9Ebjd5kPPr1Z7xar8kDcVi+TVfFjHLuc5rlr/o2GO7bbQuTYsJSVJM8H4wU4Ibxd/z+bPS7efn9bac7oPwSV3Wph0z9gYgOlBc4ow9I4D6aR4UoHrd2c1rk4SM31jG7d0fU/5tQYfAPJtRgAQjpJB88sJ0mOgnefPcs5zx7ppIh84ZmY4DFeyPz/sFYGH4wqjfRSP/xOKc2u6FsbSdV8aLee9aDsyVX5bPebsJwS3bBJizL36ZTvWM=
```

**Comment by Yousei9** [comment]
Thanks for the extensive report, however I'm not sure what you want of me. LTN only acts according to the signals you feed into the station.

I built stations with triggered requests myself and never ran into such problems so I have to assume your circuit logic is faulty.
If it's only 1-3 signals requiring higher request volumes I'll use my Blacklist filter for all but those and handle them in dedicated deciders parallel to the filter.
This one for example triggers lubricant at a much higher volume than the items: 
`0eNq1WduOozgQ/Rc/rvAIG3NJNJqv2LdVFDngTqwBwxgTTdTi39eGDnESLibd3Q+tEPBxuU5VnSryDg55wyrJhQLbd8DTUtRg+987qPlR0Nx8py4VA1vAFSuABwQtzJWkPAetB7jI2F+wRe3OA0worjjr13cXl71oigOT+oFhpdlBUaFgWhYHLqgqpYatylqvLYXZUOPB4EfogYv+gH6EepuMS5b29yPPQChZ5vsDO9Ez1+v1ojeeKyadbGe5xpKl4ClMuUwbrsxJ0rIxPoCRr/+Gg8Wtt4hHszMVKcuW0RIHtEqWKatrLo6wEfdg5A5s4wB2oEp75TINgnwHlLqiKYN1ypk+JtQXv2cAkQNgWtJ8BiJwsUmVgsGD5LPGkHbX3RSiD5/a4CHz7ygZE3ak8gxsidnYutaH2bWt9eU1nPHNjoLmOcxpUY1EMb5Gsf4wBhOszAo0mRX4haw4c6ka/c3tLN0T0LcdihGJSRJEJLGS3VuBhiw05MdBTFCCyQCGV4FhCywMoiT2NwgPWMEqrMDCwlFCgpCE0YBFVmER+5ABwSiO8c1h4Sqs0MKKYuQnSXRzV7QKKrKggiAMCQnwQmmbgortE0axPh+KFuraFFRiQSWBPp6fLBS1KaSNhUTQhgQ+sSvbbiznyABSsIw3BezlQItBVeZsLon7pHMuJuihmBjh6gWiu8SPteap+MSPC8ZrUXgTIsnVqWDKCNtMGcHrlPWGute3Mz4c/Y3LWu2dyWI0PRm+amZg9teKpyNTNwdlxSTtzQA/f+q1ZaOqZjV6u4aguBMHbO5Llj3R9+D+rsUZJ9qRqGiwOWMpz5h0YQk7svQB6UbRW95oQwdz8sboqOGid98HL7r292qqzayo7MzU7MyQM49bXfZdru7fZFnsudAQYKtkw6ZYG2ElmWRhc2NzQtvHWYnXafB3NqbjPQ1a0MrZ1gotiOOKzhgtaKNzU4wWhNG1H0YLqrjUCqNVXf5sE4wsKdy5R7N7jQmfHhwN5uTFYMZfEczz2W/15749N7p7K3LzwWalwJNlfR+j7knOp6h7MttbioHRYyHfaeYgCyMHenkS978iRp7kO1cCFvQvVJJyAXMmjupk94drBwWDJ9mfhtXaCKhOktWnMs/uRhpih2DojqtL0bnT7THYTitt4Gjd7DnJO7576QI/TJ0m6oknPIZ6Gz3z8shr0zP2FGgRqmAv8DPN8EQwvJI5kwcnsyb2/cNMu+5f0/kpQq+7D73SC90sFRd10qJ03zI9Nku/zO2r9ff7LSxqaqafz0uTSF2PtOo1xoRHwzmPjvmy59op7Ued+h2ekYxmfQvZ2d67xwPdxd4cpGKZO4//gvZTRd+s1tZ+jC5gzOvRawOa7/iO59vnM/gwn/3z5ePZmKuD2els86ii4VTMx69MXf53TF1Ljp+O+9cH4q+au8L5lxfxJ8Yyb36jCVoTa2rSmX88Kdj9JvFcxEhHKXbQxI0zJnbFxL4rpu8MiVwh3a3ErpATztTsd9PS1vo5yQNn3Qb22ZPEGCdo48e4bf8HbiXYaA==`

For more than 3 I put a whitelist filter in parallel.

All bp can be found in the design topic. https://forums.factorio.com/viewtopic.php?f=214&t=51073&start=20&hilit=whitelist

**Comment by Yousei9** [comment]
> I also now understand better that that LTN will not automatically subtract the content of a delivery to the signal it acted upon.

LTN does keep track of deliveries until it either reaches the depot, times out or the train changes. See known issues/incompatible mods for the last case.

> So why multiple deliveries generated at the same moment instead of single large one matching this same static value?

Impossible to tell without having a full debug log enabled logfile. 
On LTN side there probably was no valid provider available to provide the request amount in one shipment. Deliveries reserve their amount on creation, so other Trains might have virtually emptied the providers.
However its just as possible that your circuit logic doesn't update properly with current content and triggers multiple shipment.

**Comment by digzol** [comment]
I don't think the issue is related to incompatible mods or having other trains requesting.

Actually, I checked the logfile myself and I now understand better what you meant and also how LTN works this out.

My problem really is fully explained with this one line:
`updating requested count with train inventory: item,sulfur -14000+11826=-2174`
(static signal + train cargo = items unloaded)
Because I'm not adding the unloaded item count to the request signal, this is what happens.

I'm not sure if anything can be done about this without breaking other designs, but the design I posted ends up being faulty due to misunderstanding that unload logic. I had in mind that `request update` only happens *after* a train leaves the station and not during.
